### ISSUE TITLE
A nullable type

### ISSUE THREAD

**Issue by samhh** [unsplash/intlc#105]
Some context: https://github.com/unsplash/intlc/issues/102#issuecomment-1089919665

**Comment by samhh** [comment]
How could we prevent the following without introducing our own typechecking?:

```
{x} {x, nullable, null {} other {{x}}}
```

In the above design `x` is probably `string | null` however that makes the first interpolation unsafe.

**Comment by OliverJAsh** [comment]
@samhh In theory this could be caught by running [`@typescript-eslint/restrict-template-expressions`](https://github.com/typescript-eslint/typescript-eslint/blob/42a5309277199e8460d15bbcbeb8d0ecf16cecae/packages/eslint-plugin/docs/rules/restrict-template-expressions.md) on the output code?

**Comment by samhh** [comment]
@OliverJAsh I don't think we can leave this for userland. This'd be a type error on output:

```
{x, number} {x, nullable, null {} other {}}
```

Where `number | null` would be fed to `Intl.NumberFormat`.

**Comment by OliverJAsh** [comment]
@samhh Perhaps it's OK to have a type error on output? I guess it's a trade-off: defer validation to TS/`tsc` vs provide validation upfront.

**Comment by samhh** [comment]
I think the above would actually compile to `number` via intersection of arguments, making it safe.

```ts
string & (string | null) = string
```
